HeatPump:AirToWater followup PR update test idf

yujiex opened the pull request:
Pull request overview
---------------------

<!-- NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE -->

### Description of the purpose of this PR

This PR addresses comments from @jmarrec about the test files in https://github.com/NREL/EnergyPlus/pull/11001
- need to replace the dummy curves with more realistic data
- the description of the test idfs are not updated
- unrelated EMS code in the cooling_only test file

### Pull Request Author

<!-- Add to this list or remove from it as applicable.  This is a simple templated set of guidelines. -->

 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer

<!-- This will not be exhaustively relevant to every PR. -->

 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

Commit:
initial commit realistic curve test idf

Commit:
update description comments, remove EMS in cooling_only one, add output

Commit:
update curves in heating and cooling_only models

jmarrec committed:
missing \type choice

jmarrec commented on idd/Energy+.idd.in:
```diff
@@ -76318,6 +76318,7 @@ HeatPump:AirToWater,
    A5,  \field Operating Mode Control Option for Multiple Unit
         \note this determines how the operation mode is assigned when the load
         \note is too large to be met by multiple heat pump units
+        \type choice
```
I added this missing type.

I'm currently trying to scope how I would wrap this object in the OpenStudio SDK.

I had already figured I would probably break the heating/cooling speeds into subobjects, but I might actually break the cooling and heating side into two subobjects too.

Have you considered this versus a large object (105 fields)? Just curious.


https://docs.google.com/spreadsheets/d/1uCcEdmM4p3M9L7HeihAwL9F9NawK1TR8j3ADBqS6jTs/edit?usp=sharing

Placement of field "Minimum Part Load Ratio" is strange to me, it's "stuck" between heating-related fields

jmarrec replied:
Why did you pick "min-fields 56"?

jmarrec replied:
I don't understand the booster mode for cooling/heating. Didn't find an entry in the Engineering reference.

jmarrec replied:
Essentially, this appears to just allow using 6 speeds instead of 5, but that sixth speed isn't used to determine the reference Capacity and COP (edit: this statement is apparently wrong)

jmarrec replied:
I think there's a weirdness and potential error in the way the reference capacity is set.

I/O ref says

> The maximum number of speeds allowed is 5. Speed level 1 is the lowest speed; and speed level 5 is the highest
speed.

> Rated heating capacity at speed i, where i is the speed number from 1 to 5. For speed 1, the default value is autosize.

So, it seems that the reference Capacity is taken at speed 1, and the flag `referenceCapacityWasAutoSized` is set based on that

https://github.com/NREL/EnergyPlus/blob/a4c9dbeddba8db06900cdda7834ba6ef9e299cab/src/EnergyPlus/PlantLoopHeatPumpEIR.cc#L4162-L4167

Then it's assigned to speed 5 here:

https://github.com/NREL/EnergyPlus/blob/a4c9dbeddba8db06900cdda7834ba6ef9e299cab/src/EnergyPlus/PlantLoopHeatPumpEIR.cc#L4270-L4272

jmarrec replied:
* If speed 5 is the nominal speed level, shouldn't it be the one carrying the "autosize"? And shouldn't the `thisAWHP.ratedCapacity[thisAWHP.numSpeeds - 1]` be reassigned the autosized capacity?

* It seems that if you have a booster mode, the block on line 4270 to 4272 actually takes the booster mode as a the reference capacity. is that correct?

yujiex replied:
> Have you considered this versus a large object (105 fields)? Just curious.

The intention was so that it resembles the real equipment, whose heating and cooling functionality is usually encapsulated in one piece of equipment. Although it would have been easier to implement with heating and cooling as separate component.

> Placement of field "Minimum Part Load Ratio" is strange to me, it's "stuck" between heating-related fields

Indeed. I just moved it up to before the group of heating input.

> Why did you pick "min-fields 56"?

It should be 77. To make sure the number of speeds for heating and cooling gets defaults if user didn't give inputs. I just changed it.

> I don't understand the booster mode for cooling/heating. Didn't find an entry in the Engineering reference.

To my understanding, if the booster mode is enabled, the heat pump will have the ability to increase its capacity (increase compressor speed) further beyond the maximum, to meet a large demand. But this is usually at the cost of lower COP. If it's not enabled, the heat pump will operate at the maximum speed level and not meet the large demand.

I only have description of booster mode in IO reference as follows. I can add some more description of its intention to engineering reference. 

```
Field: Booster Mode On Heating. This field is a flag that indicates
whether the heat pump has a booster mode enabled for heating operation. If the
flag is on, booster mode will be enabled and the heat pump will operate at a
higher capacity during heating operation. The default value is false, which
means booster mode is not enabled. When the booster mode is on, the following
performance fields are required.
```

> Essentially, this appears to just allow using 6 speeds instead of 5, but that sixth speed isn't used to determine the reference Capacity and COP.

Yes, it will use another set of performance curves to represent the booster mode. 
It is used to determine reference capacity and COP. Here when the booster mode is on, the curves, COP, capacity etc are added as another speed level. In the later calculation of reference COP and capacity, the booster mode data is considered as well.

```
                if (thisAWHP.boosterOn) {
                    thisAWHP.numSpeeds += 1;
```

> I think there's a weirdness and potential error in the way the reference capacity is set.

Indeed, I will modify the code to have the highest speed level be autosizable (currently it's fixed at the lowest speed) and make the autosized capacity be reference capacity.

> It seems that if you have a booster mode, the block on line 4270 to 4272 actually takes the booster mode as a the reference capacity. is that correct?

Yes, the booster mode is treated as a higher speed level and its capacity will be considered in the reference capacity

jmarrec replied:
I guess my followup question is why do you need a booster speed in that case?

Just have the potential for 6 speeds. And the Nth highest speed is hte booster one. each speed level includes the curves, COP and capacity, so I don't understand how it's different from regular speeds?

jmarrec replied:
(and thanks for the taking the time to answer the questions! I'm not trying to be difficult, this is just a byproduct of having to wrap this object in the OpenStudio SDK).

Dunno if you saw my question lower here; https://github.com/NREL/EnergyPlus/pull/11251#issuecomment-3371959915

jmarrec replied:
> The intention was so that it resembles the real equipment, whose heating and cooling functionality is usually encapsulated in one piece of equipment. Although it would have been easier to implement with heating and cooling as separate component.

FWIW, one does not prevent the other. We have a bunch of equipment that has a top-level wrapper object and logical subojects. (CentralHeatPumpSystem and ChillerHeaterPerformance:Electric:EIR to name one)

yujiex replied:
I see. Maybe that parent-child arrangement with separate heating/cooling component as child is a cleaner way.

jmarrec committed:
IDD correctness + strip trailing whitespace

jmarrec commented:
I've got another question. I'm seeing on the branch that this type...

```diff
  Branch,
    Heating Coil Load Loop Supply Branch,  !- Name
    ,                        !- Pressure Drop Curve Name
    Pump:VariableSpeed,      !- Component 1 Object Type
    Heating Coil Load Loop Pump,  !- Component 1 Name
    Heating Coil Load Loop Supply Inlet Node,  !- Component 1 Inlet Node Name
    Heating Coil Load Loop Intermediate Node,  !- Component 1 Outlet Node Name
~    HeatPump:AirToWater:Heating,  !- Component 2 Object Type
    test_AWHP,               !- Component 2 Name
    Heating Coil Load Loop Intermediate Node,  !- Component 2 Inlet Node Name
    Heating Coil Load Loop Supply Outlet Node;  !- Component 2 Outlet Node Name
```

Is this expected and documented?

Commit:
move min part load ratio input in idd out of heating region

Commit:
change min-fields

yujiex commented:
> I've got another question @yujiex . I'm seeing on the branch that this type...
> 
> ```diff
>   Branch,
>     Heating Coil Load Loop Supply Branch,  !- Name
>     ,                        !- Pressure Drop Curve Name
>     Pump:VariableSpeed,      !- Component 1 Object Type
>     Heating Coil Load Loop Pump,  !- Component 1 Name
>     Heating Coil Load Loop Supply Inlet Node,  !- Component 1 Inlet Node Name
>     Heating Coil Load Loop Intermediate Node,  !- Component 1 Outlet Node Name
> -   HeatPump:AirToWater:Heating,  !- Component 2 Object Type
>     test_AWHP,               !- Component 2 Name
>     Heating Coil Load Loop Intermediate Node,  !- Component 2 Inlet Node Name
>     Heating Coil Load Loop Supply Outlet Node;  !- Component 2 Outlet Node Name
> ```
> 
> Is this expected and documented?

this is because when there are both heating and cooling component, if they have the same type, they'll have this error: "Same component name and type has differing Node Names"

Commit:
move min part load ratio after op mode ctrl sched, update idf

Commit:
fix unit test idd change

jmarrec committed:
Copy paste error in note field

Commit:
allow autosize for max speed level capacity

Commit:
Merge remote-tracking branch 'origin/AWHPfollowup' into AWHPfollowup

Commit:
fix heating_only test, remove op mode EMS

yujiex commented:
> > this is because when there are both heating and cooling component, if they have the same type, they'll have this error: "Same component name and type has differing Node Names"
> 
> No, you can't have object types on the branch that differ from the actual object type in the IDD. That error means something is incorrect in the node or compset registrations. Lots of other equipment have the same object on different branches (e.g. chiller chilled water nodes and condenser nodes).

I see, I will investigate what is the issue and fix it.

Commit:
remove :heating :cooling in component type in branches

Commit:
autosize fix, max speed level miss offset, clang-format

yujiex commented:
@jmarrec I changed autosize to maximum speed capacity, instead of fixed it at speed 1
@mjwitte I fixed the TestCompSet and the idfs don't have :heating or :cooling in the branch object type.
Please see if they look okay. Thanks.

jmarrec committed:
Correct case for UnivariateFunctions / BivariateFunctions

OpenStudio enforces the references, that's why I found this on this object and realized more objects had the same issue.

Commit:
defrost heater capacity remove autosize

Commit:
Merge remote-tracking branch 'origin/AWHPfollowup' into AWHPfollowup

Commit:
add reference-class-name etc to the name field in idd

mitchute committed:
Merge branch 'develop' into AWHPfollowup

rraustad commented on src/EnergyPlus/PlantLoopHeatPumpEIR.cc:
```diff
@@ -4159,15 +4163,14 @@ void HeatPumpAirToWater::processInputForEIRPLHP(EnergyPlusData &state)
+                                        format("cannot autosize capacity below maximum speed (name={}, field={})",
+                                               thisAWHP.name,
+                                               format("rated_{}_capacity_at_speed_{}", modeKeyWord, i + 1)));
+                        errorsFound = true;
```
The IDD shows all speeds are autosizable. This check trips a severe if any lower speeds are autosized. Is that correct?

yujiex replied:
Yes, It can only size the max capacity, the reference capacity for intermediate speed levels will just come from the user for now. 

Should we allow lower speeds be autosized too? like maybe first figure out the max speed capacity, and evenly divided by the speed levels (if the max speed capacity is sized to 100W, and that there are two speeds, then assign the lower speed to 50W)? Or would that be more confusing?
@rraustad

rraustad replied:
Using NumSpeeds to calculate lower stage capacity is the way the VS coils autosize. The user will see that these fields are autosizable and would likely try autosizing all speeds.

<img width="478" height="180" alt="image" src="https://github.com/user-attachments/assets/48a23b5a-21de-4841-90fc-8cc3956d6481" />

mjwitte commented:
@yujiex The registered component type is not correct. It must match the actual object type in the idf/IDD.
In this table, HeatPumpAirToWaterHeating and HeatPumpAirToWaterCooling should both be HEATPUMP:AIRTOWATER. (This is from PlantLoopHeatPump_EIR_AirSource_and_AWHP).

<img width="1765" height="370" alt="image" src="https://github.com/user-attachments/assets/2bebc238-79bb-49dc-b0bb-7631c20ded5c" />

mjwitte committed:
Remove trailing spaces

mjwitte committed:
Extend various plant arrays for AWHP and earlier fuel-fired AWHP

mjwitte committed:
aargh format

yujiex commented:
> Alright, thanks all. @yujiex we'll merge this today, but please take another look at all the traffic here and submit a second follow-up PR soon.

will do. Thanks everyone.
